moab_state_transform.ink

```diff
@@ -0,0 +1,137 @@
+
+type SimState {
+    # Ball X,Y position
+    ball_x: number<-MaxDistancePerStep - RadiusOfPlate .. RadiusOfPlate + MaxDistancePerStep>,
+    ball_y: number<-MaxDistancePerStep - RadiusOfPlate .. RadiusOfPlate + MaxDistancePerStep>,
+
+    # Ball X,Y velocity
+    ball_vel_x: number<-MaxVelocity .. MaxVelocity>,
+    ball_vel_y: number<-MaxVelocity .. MaxVelocity>,
+}
+
+# State received from the simulator after each iteration
+type ObservableState {
+    # Ball X,Y position
+    ball_x: number<-MaxDistancePerStep - RadiusOfPlate .. RadiusOfPlate + MaxDistancePerStep>,
+    ball_y: number<-MaxDistancePerStep - RadiusOfPlate .. RadiusOfPlate + MaxDistancePerStep>,
+
+    # Ball X,Y velocity
+    ball_vel_x: number<-MaxVelocity .. MaxVelocity>,
+    ball_vel_y: number<-MaxVelocity .. MaxVelocity>,
+}
+
+# Action provided as output by policy and sent as
+# input to the simulator
+type SimAction {
+    # Range -1 to 1 is a scaled value that represents
+    # the full plate rotation range supported by the hardware.
+    input_pitch: number<-1 .. 1>, # rotate about x-axis
+    input_roll: number<-1 .. 1>, # rotate about y-axis
+}
+
+# Per-episode configuration that can be sent to the simulator.
+# All iterations within an episode will use the same configuration.
+type SimConfig {
+    # Model initial ball conditions
+    initial_x: number<-RadiusOfPlate .. RadiusOfPlate>, # in (m)
+    initial_y: number<-RadiusOfPlate .. RadiusOfPlate>,
+
+    # Model initial ball velocity conditions
+    initial_vel_x: number<-MaxInitialVelocity .. MaxInitialVelocity>, # in (m/s)
+    initial_vel_y: number<-MaxInitialVelocity .. MaxInitialVelocity>,
+
+    # Range -1 to 1 is a scaled value that represents
+    # the full plate rotation range supported by the hardware.
+    initial_pitch: number<-1 .. 1>,
+    initial_roll: number<-1 .. 1>,
+}
+
+function TransformState (s: SimState): ObservableState {
+    return {
+        ball_x: 0 - s.ball_x,
+        ball_y: 0 - s.ball_y,
+        ball_vel_x: 0 - s.ball_vel_x,
+        ball_vel_y: 0 - s.ball_vel_y,
+    }
+}
+
+# Define a concept graph with a single concept
+graph (input: ObservableState) {
+    concept MoveToCenter(input): SimAction {
+        curriculum {
+            # The source of training for this concept is a simulator that
+            #  - can be configured for each episode using fields defined in SimConfig,
+            #  - accepts per-iteration actions defined in SimAction, and
+            #  - outputs states with the fields defined in SimState.
+            source simulator MoabSim(Action: SimAction, Config: SimConfig): SimState {
+                # Automatically launch the simulator with this
+                # registered package name.
+                package "Moab"
+            }
+
+            training {
+                # Limit episodes to 250 iterations instead of the default 1000.
+                EpisodeIterationLimit: 250
+            }
+
+            state TransformState
```

Comment:
add a comment explaining what this does

Reply:
# Instruct the training/assessment session to use the state transform function
            # Exporting a brain will still expect the ObservableAction type